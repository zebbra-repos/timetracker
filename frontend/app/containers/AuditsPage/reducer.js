import { fromJS } from 'immutable';
import moment from 'moment-timezone';

import { SET_SELECTED_YEAR, SET_RENDER_COMPONENT } from './constants';
import {
  fetchAuditLogs,
} from './routines';

const initialState = fromJS({
  selectedYear: moment().isoWeekYear(),
  loading: false,
  audits: [],
  audit: fromJS({}),
  renderComponent: 'table',
});

/* eslint-disable no-param-reassign */
const sanitizeAudits = (audits) => audits.map((audit) => {
  try {
    audit.username = audit.user && audit.user.username;
    if (audit.arguments) {
      audit.params = audit.arguments.params;
      audit.query = audit.arguments.query;
      audit.headers = audit.arguments.headers;

      if (audit.arguments.args) {
        audit.data = audit.arguments.args.data;
      }

      delete audit.arguments;
    }
  } catch (error) {
    // eslint-disable-next-line no-console
    console.log(error);
  }
  return audit;
});

function auditsPageReducer(state = initialState, action) {
  switch (action.type) {
    case SET_SELECTED_YEAR:
      return state
        .set('selectedYear', action.year);

    case SET_RENDER_COMPONENT:
      return state
        .set('renderComponent', action.component)
        .set('audit', fromJS(action.audit));

    case fetchAuditLogs.TRIGGER:
      return state
        .set('loading', true);
    case fetchAuditLogs.FULFILL:
      return state
        .set('loading', false);
    case fetchAuditLogs.SUCCESS:
      return state
        .set('audits', fromJS(sanitizeAudits(action.payload)));

    default:
      return state;
  }
}

export default auditsPageReducer;
