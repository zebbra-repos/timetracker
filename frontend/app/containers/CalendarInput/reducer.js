/*
 *
 * CalendarInput reducer
 *
 */

import { fromJS } from 'immutable';

import { SET_INITIAL_VALUE } from './constants';

const initialState = fromJS({
  initialValue: undefined,
});

function calendarInputReducer(state = initialState, action) {
  switch (action.type) {

    // Handle set initial cell input values
    case SET_INITIAL_VALUE:
      return state
        .set('initialValue', action.value);

    default:
      return state;
  }
}

export default calendarInputReducer;
