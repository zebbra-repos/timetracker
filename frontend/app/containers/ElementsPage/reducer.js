/*
 *
 * ElementsPage reducer
 *
 */

import { fromJS } from 'immutable';
import { unionBy, filter } from 'lodash';

import { SET_RENDER_COMPONENT } from './constants';
import { fetchElements, createElement, editElement, deleteElement } from './routines';

const initialState = fromJS({
  elements: [],
  element: fromJS({}),
  renderComponent: 'table',
  loading: false,
});

/* eslint-disable no-param-reassign */
const sanitizeElements = (elements) => elements.map((element) => {
  element.start = new Date(element.start);
  if (element.end) {
    element.end = new Date(element.end);
  }
  if (element.minEnd) {
    element.minEnd = new Date(element.minEnd);
  }
  if (element.maxStart) {
    element.maxStart = new Date(element.maxStart);
  }
  return element;
});
/* eslint-enabel no-param-reassign */

function elementsPageReducer(state = initialState, action) {
  switch (action.type) {
    // Handle GET
    case fetchElements.TRIGGER:
      return state
        .set('loading', true);
    case fetchElements.FULFILL:
      return state
        .set('loading', false);
    case fetchElements.SUCCESS:
      return state
        .set('elements', fromJS(sanitizeElements(action.payload)));

    // Handle POST and PUT
    case createElement.SUCCESS:
    case editElement.SUCCESS:
      return state
        .update('elements', (data) => fromJS(unionBy(sanitizeElements([action.payload]), data.toJS(), 'id')));

    // Handle DELETE
    case deleteElement.SUCCESS:
      return state
        .update('elements', (data) => fromJS(filter(data.toJS(), (element) => element.id !== action.payload.id)));

    // Handle set render component
    case SET_RENDER_COMPONENT:
      return state
        .set('renderComponent', action.component)
        .set('element', fromJS(action.element));

    default:
      return state;
  }
}

export default elementsPageReducer;
