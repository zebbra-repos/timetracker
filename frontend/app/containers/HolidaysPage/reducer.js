/*
 *
 * HolidaysPage reducer
 *
 */

import { fromJS } from 'immutable';
import moment from 'moment-timezone';
import { unionBy, filter } from 'lodash';

import { SET_RENDER_COMPONENT, SET_SELECTED_YEAR } from './constants';
import { fetchHolidays, createHoliday, editHoliday, deleteHoliday } from './routines';

const initialState = fromJS({
  selectedYear: moment().isoWeekYear(),
  holidays: [],
  holiday: fromJS({}),
  renderComponent: 'table',
  loading: false,
});

/* eslint-disable no-param-reassign */
const sanitizeHolidays = (holidays) => holidays.map((holiday) => {
  holiday.date = new Date(holiday.date);
  return holiday;
});
/* eslint-enabel no-param-reassign */

function holidaysPageReducer(state = initialState, action) {
  switch (action.type) {
    // Handle GET
    case fetchHolidays.TRIGGER:
      return state
        .set('loading', true);
    case fetchHolidays.FULFILL:
      return state
        .set('loading', false);
    case fetchHolidays.SUCCESS:
      return state
        .set('holidays', fromJS(sanitizeHolidays(action.payload)));

    // Handle POST and APTH
    case createHoliday.SUCCESS:
    case editHoliday.SUCCESS:
      return state
        .update('holidays', (data) => fromJS(unionBy(sanitizeHolidays([action.payload]), data.toJS(), 'id')));

    // Handle DELETE
    case deleteHoliday.SUCCESS:
      return state
        .update('holidays', (data) => fromJS(filter(data.toJS(), (holiday) => holiday.id !== action.payload.id)));

    // Handle set render component
    case SET_RENDER_COMPONENT:
      return state
        .set('renderComponent', action.component)
        .set('holiday', fromJS(action.holiday));

    // Handle set selected year
    case SET_SELECTED_YEAR:
      return state
        .set('selectedYear', action.year);

    default:
      return state;
  }
}

export default holidaysPageReducer;
