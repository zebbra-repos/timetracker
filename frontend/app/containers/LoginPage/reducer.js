/*
 *
 * LoginPage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  SET_COMPONENT,
} from './constants';
import {
  login,
  verifyUser,
  requestPasswordReset,
  passwordReset,
} from './routines';

const initialState = fromJS({
  component: 'login',
  uid: undefined,
  error: undefined,
  info: undefined,
  token: undefined,
});

function loginPageReducer(state = initialState, action) {
  switch (action.type) {

    case SET_COMPONENT:
      return state
        .set('component', action.component)
        .set('uid', action.uid)
        .set('error', undefined)
        .set('info', action.info)
        .set('token', action.token);

    case login.SUCCESS:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', undefined)
        .set('info', undefined)
        .set('token', undefined);

    case login.FAILURE:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', undefined)
        .set('info', undefined)
        .set('token', undefined);

    case verifyUser.SUCCESS:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', undefined)
        .set('info', action.payload.info)
        .set('token', undefined);

    case verifyUser.FAILURE:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', 'Es gab ein Problem beim Versenden der Email. Versuche es etwas später erneut.')
        .set('info', action.payload.info)
        .set('token', undefined);

    case requestPasswordReset.SUCCESS:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', undefined)
        .set('info', 'Passwort-Reset Email wurde versendet.')
        .set('token', undefined);

    case requestPasswordReset.FAILURE:
      return state
        .set('uid', undefined)
        .set('error', action.payload.error === 'not found' ? 'Email-Adresse wurde nicht gefunden' : 'Es gab ein Problem beim Versenden der Email. Versuche es etwas später erneut.')
        .set('info', undefined)
        .set('token', undefined);

    case passwordReset.SUCCESS:
      return state
        .set('component', 'login')
        .set('uid', undefined)
        .set('error', undefined)
        .set('info', 'Passwort erfolgreich zurückgesetzt.')
        .set('token', undefined);

    case passwordReset.FAILURE:
      return state
        .set('uid', undefined)
        .set('error', 'Es gab ein Problem beim Ändern des Passwortes. Versuche es etwas später erneut.')
        .set('info', undefined);

    default:
      return state;
  }
}

export default loginPageReducer;
