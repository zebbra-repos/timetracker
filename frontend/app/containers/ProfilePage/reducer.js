/*
 *
 * ProfilePage reducer
 *
 */

import { fromJS } from 'immutable';

import { fetchProfile, editProfile } from './routines';

const initialState = fromJS({
  username: undefined,
  email: undefined,
  firstName: undefined,
  lastName: undefined,
  roles: undefined,
});

function profilePageReducer(state = initialState, action) {
  switch (action.type) {

    // Handle GET profile
    case fetchProfile.SUCCESS:
      return fromJS(action.payload);

    case fetchProfile.FAILURE:
      return initialState;

    // Handle PUT profile
    case editProfile.SUCCESS:
      return state
        .set('firstName', action.payload.firstName)
        .set('lastName', action.payload.lastName);

    default:
      return state;
  }
}

export default profilePageReducer;
