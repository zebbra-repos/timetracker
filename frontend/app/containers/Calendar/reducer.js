/*
 *
 * Calendar reducer
 *
 */

import { fromJS } from 'immutable';
import { flatten, filter, unionBy, head, groupBy, values, mapKeys } from 'lodash';
import moment from 'moment';

import { submitCalendarForm } from 'containers/CalendarInput/routines';
import { prepareElementsReducer, navigate, initializeUserSettings } from './helpers';

import { SELECT_DATE, SET_NAVIGATION, NAVIGATE } from './constants';
import { fetchAllData, editUserSettings, deleteUserSettings } from './routines';

const initialState = () => {
  const state = fromJS({
    selectedCell: {
      elementId: undefined,
      dayOfWeek: undefined,
      rowIndex: undefined,
    },
    selectedDate: moment(),
    navigation: undefined,
    daysPerWeek: 5,
    showWeekendWithTrack: false,
    holidays: [],
    tracks: {},
    regions: [
      { type: 'range' },
      { type: 'dynamic' },
      { type: 'static' },
    ],
    elements: [],
    navigationElements: [],
    userSettings: {
      showWeekend: false,
      elements: {},
    },
    reporting: {},
    loading: false,
  });

  return state;
};

/**
 * Calculates the new selected calendar cell based on the current redux-store state of the calendar.
 * You can pass an array of elements and the number of days for which to calculate the navigation
 * if they haven changed since the last store update.
 *
 * @param {immutable map} state         The current redux-store state of the calendar container
 * @param {array}         elements      The elements for which to calculate the navigation (optional)
 * @param {number}        daysPerWeek   The number of days for which to caluclate the navigation (optional)
 */
const calculateNavigation = (state, elements, daysPerWeek) => navigate(
  elements || state.get('elements').toJS(),
  daysPerWeek || state.get('daysPerWeek'),
  state.get('selectedCell').toJS(),
  state.get('navigation').toJS(),
);

/**
 * Calculates the new date based on the current selected date and the changeWeek indicator
 *
 * @param {date}    selectedDate    The current selected date
 * @param {string}  changeWeek      Can be left or right
 */
const calculateWeek = (selectedDate, changeWeek) => {
  if (changeWeek) {
    if (changeWeek === 'left') {
      return selectedDate.clone().subtract(1, 'week');
    } else if (changeWeek === 'right') {
      return selectedDate.clone().add(1, 'week');
    }
  }
  return selectedDate;
};

function calendarReducer(state = initialState(), action) {
  let elements = null;
  let elementId = null;
  let dayOfWeek = null;
  let changeWeek = null;
  let rowIndex = null;
  let showWeekend = null;
  let showWeekendWithTrack = false;
  let userSettings = null;
  let selectedCell = null;
  let daysPerWeek = 5;
  let key = null;
  let weekday = null;

  switch (action.type) {
    // Handle select date
    case SELECT_DATE:
      return state
        .update('selectedDate', (date) => date.isSame(action.date, 'day') ? date : action.date);

    //  Navigation Handling
    case SET_NAVIGATION:
      return state
        .set('navigation', fromJS({ key: action.key, shiftKey: action.shiftKey }));

    case NAVIGATE:
      // apply the navigation stored in the state
      if (action.fromStore) {
        ({ elementId, dayOfWeek, changeWeek, rowIndex } = calculateNavigation(state));

        // if the week changed, then we only update the selectedDate
        if (changeWeek) {
          return state
            .update('selectedDate', (date) => calculateWeek(date, changeWeek));
        }
      } else {
        ({ elementId, dayOfWeek, rowIndex } = action);
      }

      // otherwise we change the selectedCell
      return state
        .set('selectedCell', fromJS({ elementId, dayOfWeek, rowIndex }))
        .set('navigation', undefined);

    // Handle fetch all data
    case fetchAllData.TRIGGER:
      return state
        .set('loading', true);
    case fetchAllData.FULFILL:
      return state
        .set('loading', false);
    case fetchAllData.SUCCESS:
      // if the user does not have any custom settings for the calendar component
      // then we initialize with the default user settings
      userSettings = head(action.payload.settings);
      if (!userSettings) {
        userSettings = initializeUserSettings(action.payload.elements);
      }

      ({ elements, showWeekend, showWeekendWithTrack } = prepareElementsReducer(
        action.payload.elements,
        action.payload.tracks,
        action.payload.holidays,
        userSettings,
        state.get('selectedDate')
      ));

      // decide how many days per week to display
      daysPerWeek = (showWeekend || showWeekendWithTrack) ? 7 : 5;

      // if there is pending navigation open (in case of week-change) then
      // we apply the navigation here
      if (state.get('navigation')) {
        ({ elementId, dayOfWeek, rowIndex } = calculateNavigation(state, elements, daysPerWeek));
        selectedCell = fromJS({
          elementId,
          dayOfWeek,
          rowIndex,
        });
      } else {
        selectedCell = state.get('selectedCell');
      }

      return state
        .set('selectedCell', selectedCell)
        .set('navigation', undefined)
        .set('userSettings', fromJS(userSettings))
        .set('daysPerWeek', daysPerWeek)
        .set('showWeekendWithTrack', showWeekendWithTrack)
        .set('holidays', fromJS(action.payload.holidays))
        .set('elements', fromJS(elements))
        .set('tracks', fromJS(groupBy(action.payload.tracks, (track) => moment(track.date).weekday())))
        .set('reporting', fromJS(mapKeys(action.payload.reporting, (value, weekdayKey) => moment(weekdayKey, 'YYYYMMDD').weekday())));

    // Handle calendar form submit routine
    case submitCalendarForm.SUCCESS:
      key = Object.keys(action.payload.reporting)[0];
      weekday = moment(key, 'YYYYMMDD').weekday();

      switch (action.payload.action) {
        case 'DELETE':
          return state
            .updateIn(['reporting', weekday.toString()], () => fromJS(action.payload.reporting[key]))
            .updateIn(['tracks', weekday.toString()], (data) => fromJS(filter(((data && data.toJS()) || []), (track) => track.id !== action.payload.id)));
        case 'POST':
        case 'PUT':
          return state
            .updateIn(['reporting', weekday.toString()], () => fromJS(action.payload.reporting[key]))
            .updateIn(['tracks', weekday.toString()], (data) => fromJS(unionBy(action.payload.tracks, (data && data.toJS()) || [], 'id')));

        default: return state;
      }

    // handle edit user settings routine
    case editUserSettings.TRIGGER:
      return state
        .set('loading', true);
    case editUserSettings.FULFILL:
      return state
        .set('loading', false);
    case editUserSettings.SUCCESS:
      ({ elements, showWeekend, showWeekendWithTrack } = prepareElementsReducer(
        state.get('elements').toJS(),
        flatten(values(state.get('tracks').toJS())),
        state.get('holidays').toJS(),
        action.payload,
        state.get('selectedDate')
      ));

      return state
        .set('userSettings', fromJS(action.payload))
        .set('daysPerWeek', showWeekend ? 7 : 5)
        .set('showWeekendWithTrack', showWeekendWithTrack)
        .set('elements', fromJS(elements));

    case deleteUserSettings.TRIGGER:
      return state
        .set('loading', true);
    case deleteUserSettings.FULFILL:
      return state
        .set('loading', false);
    case deleteUserSettings.SUCCESS:
      ({ elements, showWeekend, showWeekendWithTrack } = prepareElementsReducer(
        state.get('elements').toJS(),
        flatten(values(state.get('tracks').toJS())),
        state.get('holidays').toJS(),
        {
          showWeekend: false,
          elements: {},
        },
        state.get('selectedDate')
      ));

      return state
        .set('userSettings', fromJS({
          showWeekend: false,
          elements: {},
        }))
        .set('daysPerWeek', showWeekend ? 7 : 5)
        .set('showWeekendWithTrack', showWeekendWithTrack)
        .set('elements', fromJS(elements));

    default:
      return state;
  }
}

export default calendarReducer;
