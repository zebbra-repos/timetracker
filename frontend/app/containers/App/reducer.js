/*
 * AppReducer
 *
 */

import { fromJS } from 'immutable';
import config from 'config';
import { LOCATION_CHANGE } from 'react-router-redux';

import { NAVIGATE } from 'containers/Calendar/constants';
import { fetchVersion } from './routines';

import {
  TOGGLE_SIDEBAR_LEFT,
  TOGGLE_SIDEBAR_RIGHT,
  SHOW_APP_ERROR,
  HIDE_APP_ERROR,
  SHOW_APP_INFO,
  HIDE_APP_INFO,
  SET_ONBOARDED,
  TOGGLE_PRINTING,
  RELOAD_AND_UPDATE_VERSION,
} from './constants';

import Pack from '../../../package.json';

// set the api url based on environment
const url = config.api.host;

// The initial state of the App
const initialState = fromJS({
  version: Pack.version,
  latestVersion: Pack.version,
  api: {
    url,
  },
  sidebarLeft: {
    show: false,
  },
  sidebarRight: {
    show: false,
    name: undefined,
  },
  error: {},
  info: {},
  onboarded: true,
  printing: false,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    // Location change
    case LOCATION_CHANGE:
    case NAVIGATE:
      return state
        .set('sidebarRight', fromJS({ show: false, name: undefined }));

    // App error handling
    case SHOW_APP_ERROR:
      return state
        .set('error', fromJS(action.error));
    case HIDE_APP_ERROR:
      return state
        .set('error', fromJS({}));

    // App info handling
    case SHOW_APP_INFO:
      return state
        .set('info', fromJS(action.info));
    case HIDE_APP_INFO:
      return state
        .set('info', fromJS({}));

    // Sidebar handling
    case TOGGLE_SIDEBAR_LEFT:
      return state
        .updateIn(['sidebarLeft', 'show'], (show) => !show);
    case TOGGLE_SIDEBAR_RIGHT:
      return state
        .updateIn(['sidebarRight', 'show'], (show) => !show)
        .setIn(['sidebarRight', 'name'], action.name);

    // Handle onboarded
    case SET_ONBOARDED:
      return state
        .set('onboarded', action.onboarded);

    // Handle printing
    case TOGGLE_PRINTING:
      return state
        .set('printing', !state.get('printing'));

    case fetchVersion.SUCCESS:
      return state
        .set('latestVersion', action.payload.version);

    case RELOAD_AND_UPDATE_VERSION:
      return state
        .set('version', state.get('latestVersion'));

    default:
      return state;
  }
}

export default appReducer;
