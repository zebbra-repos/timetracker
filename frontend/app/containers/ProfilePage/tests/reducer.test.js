import reducer from '../reducer';

describe('profilePageReducer', () => {
  it('returns the initial state', () => {
    const profilePageReducer = reducer(undefined, {});
    expect(profilePageReducer).toMatchSnapshot();
  });
});
