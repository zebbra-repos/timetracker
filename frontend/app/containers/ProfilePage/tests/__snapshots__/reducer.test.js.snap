exports[`profilePageReducer returns the initial state 1`] = `
Object {
  "email": undefined,
  "firstName": undefined,
  "lastName": undefined,
  "roles": undefined,
  "username": undefined,
}
`;
